substrate/primitives/blockchain/src/backend.rs

```diff
@@ -273,13 +293,22 @@ pub trait Backend<Block: BlockT>:
 			displaced_leaves: Vec::with_capacity(leaves.len()),
 			displaced_blocks: Vec::with_capacity(leaves.len()),
 		};
+
 		let mut displaced_blocks_candidates = Vec::new();
 
 		for leaf_hash in leaves {
 			let mut current_header_metadata =
 				MinimalBlockMetadata::from(&self.header_metadata(leaf_hash)?);
 			let leaf_number = current_header_metadata.number;
 
+			if leaf_number == Zero::zero() {
+				debug!(target: crate::LOG_TARGET, %leaf_hash, "Leaf is genesis hash, removing.");
+				if finalized_block_number > leaf_number {
+					result.displaced_leaves.push((leaf_number, leaf_hash));
```

Comment:
Shouldn't it also be added to `displaced_blocks`?

Reply:
Hmm so my idea here was that displaced_leaves is for removing entries from the `leaves`, which is what we want. It is however not really a displaced branch where the genesis block itself needs to be pruned.

Reply:
Then this needs to be delayed just like everything else until the gap is closed, right?

Reply:
This is again going back to my other topic that this shouldn't assume gap will necessarly be closed at any point in the future. This genesis block may end up being pruned instead, but we need to be able to know that it is a detached leaf in order to be able to do that.

Reply:
You are right, the correct generic way of handling this would be to delay the genesis block too. It comes however at the cost of performance of this method. If we don't remove the genesis block here, we will always have more than 2 leaves until the block gap is closed (if ever, as you say). So while the block gap is not closed, we will always iterate from the block-to-be-finalized to the first block after the gap. So by removing this genesis leaf here, we can avoid this iteration at the cost of leaking at most the genesis block, if the gap is never closed.

However, if we had the pruning handling of before-gap blocks outside the pruning window you described in the other comment, we would not have this problem. Currently I am leaning in favor of removing the genesis leaf handling "hack" here and implementing the pruning improvements rather soon.

Reply:
Well, it is not up to me to decide, but for completeness after each individual PR I'd probably still not mark genesis block as displaced. Either way I think we're on the same page here and if the change happens shortly afterwards then it is fine with me as well, just want to make sure it does happen.